server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java

```diff
@@ -230,4 +228,71 @@ protected void checkDuplicates(
     updatedDuplicateOptions.putAll(customDuplicateOptions);
     checkDuplicateEntries(ctx, foundEntries, updatedDuplicateOptions);
   }
+
+  /**
+   * Flags errors for rule lists passed as parameters if there are multiple instances of mutually
+   * exclusive options.
+   *
+   * @param options Options checked to insert into error message
+   * @param rules Lists of TerminalNode to iterate through
+   * @return Number of TerminalNode instances found
+   */
+  protected int checkHasMutuallyExclusiveOptions(String options, List<TerminalNode>... rules) {
+    String tokenType = null;
+    int occurances = 0;
+    for (List<TerminalNode> option : rules) {
+      if (option.isEmpty()) continue;
+      for (TerminalNode rule : option) {
+        if (rule == null) continue;
+        if (tokenType == null) {
+          tokenType = option.get(0).getText().toUpperCase();
+          occurances++;
+          continue;
+        }
+        // Flag only options that do NOT match the first option encountered.
+        // If the option is the same then this is already flagged by checkDuplicates
+        if (!tokenType.equals(rule.getText().toUpperCase())) {
+          throwException(
+              ErrorSeverity.ERROR,
+              getLocality(rule),
+              "Exactly one option required, options are mutually exclusive: ",
+              options);
+        }
+        occurances++;
+      }
+    }
+    return occurances;
+  }
+
+  protected <E extends ParseTree> void checkHasExactlyOneOption(
+      String options, ParserRuleContext parentCtx, List<E>... rules) {
+    List<TerminalNode> children = new ArrayList<>();
+    for (List<E> rule : rules) {
+      if (!rule.isEmpty()) {
+        if (TerminalNode.class.isAssignableFrom(rule.get(0).getClass()))
+          children.addAll((List<TerminalNode>) rule);
+        else {
+          for (E context : rule) {
+            getAllTokenChildren((ParserRuleContext) context, children);
+          }
+        }
+      }
+    }
+    if (checkHasMutuallyExclusiveOptions(options, children) == 0) {
+      throwException(
+          ErrorSeverity.ERROR,
+          getLocality(parentCtx),
+          "Exactly one option required, none provided: ",
+          options);
+    }
+  }
+
+  private void getAllTokenChildren(ParserRuleContext ctx, List<TerminalNode> children) {
+    for (ParseTree child : ctx.children) {
+      if (baseDuplicateOptions.containsKey(child.getText().toUpperCase()))
```

Comment:
I suggest checking for the TerminalNode base as well, both to avoid the risk of runtime cast failures, but also to avoid stringifying possible large contexts.